blackbird_python/blackbird/utils.py

```diff
@@ -0,0 +1,245 @@
+"""
+from collections import namedtuple
+
+import numpy as np
+import sympy as sym
+from sympy.solvers import solve
+
+import networkx as nx
+from networkx.algorithms import isomorphism
+
+
+from .listener import RegRefTransform
+
+
+Command = namedtuple('Command', ['name', 'args', 'kwargs', 'modes'])
+"""namedtuple: encapsulate a specific quantum operation"""
+
+
+class TemplateError(Exception):
+    """Exception class for template related errors"""
+    pass
+
+
+def to_DiGraph(program):
+    """Convert a Blackbird program to a directed acyclic graph.
+
+    The resulting graph has nodes representing quantum operations,
+    and edges representing dependent/successor operations.
+
+    Each node is labelled by an integer; note attributes are used
+    to store information about the node:
+
+    * ``'name'`` *(str)*: name of the quantum operation (e.g., ``'S2gate'``)
+
+    * ``'args'`` *(list)*: positional arguments of the operation
+
+    * ``'kwargs'`` *(dict)*: keyword arguments of the operation
+
+    * ``'modes'`` *(tuple[int])*: modes the operation acts on
+
+    Args:
+        program (blackbird.Program): a Blackbird program
+
+    Returns:
+        networkx.DiGraph: the directed acyclic graph representing
+        the quantum program
+    """
+    grid = {}
+
+    for idx, op in enumerate(program.operations):
+        dependencies = set(op['modes'])
+
+        if 'args' in op:
+
+            for a in op['args']:
+                if isinstance(a, RegRefTransform):
+                    dependencies |= set(a.regrefs)
+
+            for _, v in op['kwargs'].items():
+                if isinstance(v, RegRefTransform):
+                    dependencies |= set(v.regrefs)
+        else:
+            op['args'] = []
+            op['kwargs'] = {}
+
+        cmd = Command(name=op['op'], args=op['args'], kwargs=op['kwargs'], modes=tuple(op['modes']))
+
+        for q in dependencies:
+            # Add cmd to the grid to the end of the line r.ind.
+            if q not in grid:
+                # add a new line to the circuit
+                grid[q] = []
+
+            grid[q].append([idx, cmd])
+
+    G = nx.DiGraph()
+
+    for q, cmds in grid.items():
+        if cmds:
+            # add the first operation on the wire that does not depend on anything
+            attrs = cmds[0][1]._asdict()
+            G.add_node(cmds[0][0], **attrs)
+
+        for i in range(1, len(cmds)):
+            # add the edge between the operations, and the operation nodes themselves
+            if cmds[i][0] not in G:
+                attrs = cmds[i][1]._asdict()
+                G.add_node(cmds[i][0], **attrs)
+
+            G.add_edge(cmds[i-1][0], cmds[i][0])
+
+    return G
+
+
+def match_template(template, program):
+    """Match a template against a Blackbird program, returning template
+    parameter values.
+
+    For example, consider the following template and program:
+
+    .. code-block:: python
+
+        template = blackbird.loads(\"\"\"\\

```

Comment:
how does this render in sphinx?

Reply:
Correctly 🙂 (i.e., `blackbird.loads("""\`). Took me a while to work out what needed to be escaped.